contracts/governance/InterchainMultisig.sol

```diff
@@ -0,0 +1,149 @@
+import { Caller } from '../utils/Caller.sol';
+import { BaseWeightedMultisig } from './BaseWeightedMultisig.sol';
+
+/**
+ * @title InterchainMultisig Contract
+ * @notice Weighted Multisig executor to call functions on any contract
+ */
+contract InterchainMultisig is Caller, BaseWeightedMultisig, IInterchainMultisig {
+    // keccak256('InterchainMultisig.Storage')
+    bytes32 internal constant INTERCHAIN_MULTISIG_STORAGE =
+        0x5a9dc2248a56f285e6221da431581a5990380ebfa07727563571b2be1517a23e;
+
+    using SafeNativeTransfer for address;
+
+    bytes32 public immutable chainNameHash;
+
+    struct InterchainMultisigStorage {
+        mapping(bytes32 => bool) isBatchExecuted;
+    }
+
+    /**
+     * @notice Contract constructor
+     * @dev Sets the initial list of signers and corresponding threshold.
+     * @param chainName The name of the chain
+     * @param weightedSigners The weighted signers payload
+     */
+    constructor(string memory chainName, WeightedSigners memory weightedSigners) BaseWeightedMultisig(0) {
+        if (bytes(chainName).length == 0) revert InvalidChainName();
+
+        chainNameHash = keccak256(bytes(chainName));
+
+        _rotateSigners(weightedSigners);
+    }
+
+    modifier onlySelf() {
+        if (msg.sender != address(this)) revert NotSelf();
+
+        _;
+    }
+
+    /**
+     * @notice Checks if a payload has been executed
+     * @param batchId The hash of the payload payload
+     * @return True if the payload has been executed
+     */
+    function isBatchExecuted(bytes32 batchId) external view returns (bool) {
+        return _interchainMultisigStorage().isBatchExecuted[batchId];
+    }
+
+    /**
+     * @notice Executes an external contract call.
+     * @notice This function is protected by the onlySigners requirement.
+     * @dev Executes a batch of calls with specified target addresses, calldata and native value.
+     * @param batchId The batchId of the multisig
+     * @param calls The batch of calls to execute
+     * @param proof The multisig proof data
+     * @dev The proof data should have signers, weights, threshold and signatures encoded
+     * @dev The signers and signatures should be sorted by signer address in ascending order
+     */
+    function executeCalls(
+        bytes32 batchId,
+        Call[] calldata calls,
+        bytes calldata proof
+    ) external payable {
+        InterchainMultisigStorage storage slot = _interchainMultisigStorage();
+        bytes32 messageHash = ECDSA.toEthSignedMessageHash(keccak256(abi.encode(batchId, calls)));
+        uint256 length = calls.length;
+
+        validateProof(messageHash, proof);
+
+        if (slot.isBatchExecuted[batchId]) revert AlreadyExecuted();
+        slot.isBatchExecuted[batchId] = true;
+
+        emit BatchExecuted(batchId, messageHash, length);
+
+        uint256 executedCalls = 0;
+
+        for (uint256 i; i < length; ++i) {
+            Call memory call = calls[i];
```

Comment:
```suggestion
            Call calldata call = calls[i];
```
Switch to calldata since we aren't modifying any variables, and so we also don't need to change `Caller` contract